File: Algorithm_LevelOne/Max_Area_of_Island/Max_Area_of_Island.py
# 695. Max Area of Island Python

class Solution(object):
    def maxAreaOfIsland(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        if not grid:
            return 0
        
        largest = 0
        row = len(grid)
        col = len(grid[0])
        
        for r in range(row):
            for c in range(col):
                if grid[r][c] == 1:
                    largest = max(largest, self.areaOfIsland(grid, r, c))
        return largest
    def areaOfIsland(self, grid, i, j):
        if i < 0 or i >= len(grid) or j < 0 or j >= len(grid[0]) or grid[i][j] != 1:
            return 0
        grid[i][j] = 0
        size = 1
        size += self.areaOfIsland(grid, i-1, j)
        size += self.areaOfIsland(grid, i+1, j)
        size += self.areaOfIsland(grid, i, j-1)
        size += self.areaOfIsland(grid, i, j+1)
        return size

File: Algorithm_LevelOne/Max_Area_of_Island/Max_Area_of_Island.cc
// 695. Max Area of Island C++
#include <vector>

class Solution {
public:
    int row;
    int col;
    int maxAreaOfIsland(std::vector<std::vector<int>>& grid) {
        if (grid.empty()) return 0;
        row = grid.size();
        col = grid[0].size();
        int largest = 0;
        for (int i = 0; i < row; i++) {
            for (int j = 0; j < col; j++) {
                if (grid[i][j] == 1) {
                    largest = std::max(largest, areaOfIsland(grid, i, j));
                }
            }
        }
        return largest;
    }
    int areaOfIsland(std::vector<std::vector<int>>& grid, int i, int j) {
        if (i < 0 || i >= row || j < 0 || j >= col || grid[i][j] != 1) return 0;
        grid[i][j] = 0;
        int size = 1;
        size += areaOfIsland(grid, i-1, j);
        size += areaOfIsland(grid, i+1, j);
        size += areaOfIsland(grid, i, j-1);
        size += areaOfIsland(grid, i, j+1);
        return size;
    }
};
